x/emissions/migrations/v3/migrate_test.go

```diff
@@ -167,5 +170,107 @@ func (s *MigrationTestSuite) TestMigrate() {
 	s.Require().True(paramsExpected.EpsilonSafeDiv.Equal(params.EpsilonSafeDiv), "%s!=%s", paramsExpected.EpsilonSafeDiv, params.EpsilonSafeDiv)
 	s.Require().True(paramsExpected.DataSendingFee.Equal(params.DataSendingFee), "%s!=%s", paramsExpected.DataSendingFee, params.DataSendingFee)
 	s.Require().Equal(paramsExpected.MaxElementsPerForecast, params.MaxElementsPerForecast)
+	s.Require().Equal(paramsExpected.MaxActiveTopicsPerBlock, params.MaxActiveTopicsPerBlock)
 	s.Require().Equal(paramsExpected, params)
 }
+
+func (s *MigrationTestSuite) TestActiveTopicsMigration() {
+	storageService := s.emissionsKeeper.GetStorageService()
+	store := runtime.KVStoreAdapter(storageService.OpenKVStore(s.ctx))
+
+	err := s.emissionsKeeper.SetParams(s.ctx, types.Params{
+		MinTopicWeight:          alloraMath.MustNewDecFromString("0"),
+		MaxActiveTopicsPerBlock: uint64(4),
+	})
+	s.Require().NoError(err)
+
+	topicCnt := 10
+	s.setUpOldTopicsData(store, topicCnt)
+	err = migrations.MigrateActiveTopics(store, s.ctx, *s.emissionsKeeper)
+	s.Require().NoError(err)
+
+	blockToActiveStore := prefix.NewStore(store, types.BlockToActiveTopicsKey)
+	iterator := blockToActiveStore.Iterator(nil, nil)
+	for ; iterator.Valid(); iterator.Next() {
+		var msg types.TopicIds
+		err := proto.Unmarshal(iterator.Value(), &msg)
+		if err != nil {
+			continue
+		}
+		s.Require().GreaterOrEqual(len(msg.TopicIds), 3)
+	}
+}
+
+func (s *MigrationTestSuite) TestLimitedActiveTopicsMigration() {
+	storageService := s.emissionsKeeper.GetStorageService()
+	store := runtime.KVStoreAdapter(storageService.OpenKVStore(s.ctx))
+
+	maxActiveTopicPerBlock := 2
+	err := s.emissionsKeeper.SetParams(s.ctx, types.Params{
+		MinTopicWeight:          alloraMath.MustNewDecFromString("0"),
+		MaxActiveTopicsPerBlock: uint64(maxActiveTopicPerBlock),
+	})
+	s.Require().NoError(err)
+
+	topicCnt := 10
+	s.setUpOldTopicsData(store, topicCnt)
+	err = migrations.MigrateActiveTopics(store, s.ctx, *s.emissionsKeeper)
+	s.Require().NoError(err)
+
+	blockToActiveStore := prefix.NewStore(store, types.BlockToActiveTopicsKey)
+	iterator := blockToActiveStore.Iterator(nil, nil)
+	for ; iterator.Valid(); iterator.Next() {
+		var msg types.TopicIds
+		err := proto.Unmarshal(iterator.Value(), &msg)
+		if err != nil {
+			continue
+		}
+		if len(msg.TopicIds) == 0 {
+			continue
+		}
+		s.Require().Len(msg.TopicIds, 3)
```

Comment:
1. Should check for errors: `s.Require().NoError(err)`
2.This `len(msg.TopicIds) == 0` validation is causing a false positive; when you comment it out, the test fails.